Source/JavaScriptCore/llint/WebAssembly.asm

```diff
@@ -1024,6 +1071,149 @@ end
         end)
 end
 
+macro slowPathForWasmTailCall(ctx, slowPath, storeWasmInstance)
+    callWasmCallSlowPath(
+        slowPath,
+        # callee is r0 and targetWasmInstance is r1
+        macro (callee, targetWasmInstance)
+            move callee, ws0
+
+            loadi ArgumentCountIncludingThis + TagOffset[cfr], PC
+
+            # the call might throw (e.g. indirect call with bad signature)
+            btpz targetWasmInstance, .throw
+
+            wgetu(ctx, m_stackOffset, ws1)
+            muli SlotSize, ws1
+
+if ARMv7
+            subp cfr, ws1, ws1
+            move ws1, sp
+else
+            subp cfr, ws1, sp
+end
+
+            storeWasmInstance(targetWasmInstance)
+            reloadMemoryRegistersFromInstance(targetWasmInstance, wa0, wa1)
+
+            wgetu(ctx, m_numberOfCalleeStackArgs, ws1)
+
+if ARM64 or ARM64E
+            leap [sp, ws1, 8], ws1
+            forEachArgumentJSR(macro (offset, gpr1, gpr2)
+                loadpairq CallFrameHeaderSize + 8 + offset[ws1], gpr1, gpr2
+            end)
+elsif JSVALUE64
+            forEachArgumentJSR(macro (offset, gpr)
+                loadq CallFrameHeaderSize + 8 + offset[sp, ws1, 8], gpr
+            end)
+else
+            forEachArgumentJSR(macro (offset, gprMsw, gpLsw)
+                load2ia CallFrameHeaderSize + 8 + offset[sp, ws1, 8], gpLsw, gprMsw
+            end)
+end
+if ARM64 or ARM64E
+            forEachArgumentFPR(macro (offset, fpr1, fpr2)
+                loadpaird CallFrameHeaderSize + 8 + offset[ws1], fpr1, fpr2
+            end)
+else
+            forEachArgumentFPR(macro (offset, fpr)
+                loadd CallFrameHeaderSize + 8 + offset[sp, ws1, 8], fpr
+            end)
+end
+
+            # We need PC to load information from context
+            # and two scratch registers to perform the copy
+            preserveGPRsUsedByTailCall(wa0, wa1)
+            move PC, ws1
+            loadi ArgumentCountIncludingThis + TagOffset[cfr], PC
+
+            # old stack size
+            wgetu(ctx, m_numberOfCallerStackArgs, wa0)
+            muli SlotSize, wa0
+            addi CallFrameHeaderSize + StackAlignment - 1, wa0
+            andi ~StackAlignmentMask, wa0
+
+            # new stack size
+            wgetu(ctx, m_numberOfCalleeStackArgs, wa1)
+            muli SlotSize, wa1
+            addi CallFrameHeaderSize + StackAlignment - 1, wa1
+            andi ~StackAlignmentMask, wa1
+
+            # new stack pointer
+            addp cfr, wa0
+            subp wa1 , wa0
+
+            move ws1, PC
+            preserveReturnAddress(ws1)
+            usePreviousFrame()
+
+if ARMv7
+            addi SlotSize, wa1
+end
+            .copyLoop:
+if ARM64 and not ADDRESS64
```

Comment:
do we need this?

Reply:
This should probably be using if JSVALUE64

Reply:
I'll have to check, I followed the semantics of `prepareForTailCall` in the JS interpreter.